In [1]:
import numpy as np
import pandas as pd
from pandas import json_normalize
import ast

pd.set_option('display.max_colwidth', None)

Read the datasets into a pandas df

In [2]:
charging_sessions_orig = pd.read_csv("charging_sessions.csv")
weather_orig = pd.read_csv("weather_burbank_airport.csv")

Start with the charging sessions, lets get some general information about the dataset.

In [3]:
charging_sessions_orig.info()
charging_sessions_orig.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66450 entries, 0 to 66449
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        66450 non-null  int64  
 1   id                66450 non-null  object 
 2   connectionTime    66450 non-null  object 
 3   disconnectTime    66450 non-null  object 
 4   doneChargingTime  62362 non-null  object 
 5   kWhDelivered      66450 non-null  float64
 6   sessionID         66450 non-null  object 
 7   siteID            66450 non-null  int64  
 8   spaceID           66450 non-null  object 
 9   stationID         66450 non-null  object 
 10  timezone          66450 non-null  object 
 11  userID            49187 non-null  float64
 12  userInputs        49187 non-null  object 
dtypes: float64(2), int64(2), object(9)
memory usage: 6.6+ MB


Unnamed: 0  kWhDelivered        siteID        userID
count  66450.000000  66450.000000  66450.000000  49187.000000
mean    4726.102062     11.787916      1.472656   1800.494582
std     3568.025394     10.314789      0.499256   2617.548614
min        0.000000      0.501000      1.000000      1.000000
25%     1845.000000      5.097000      1.000000    431.000000
50%     3960.000000      9.140000      1.000000    818.000000
75%     7027.000000     14.183000      2.000000   1805.000000
max    15291.000000    108.797242      2.000000  19923.000000

In [4]:
charging_sessions_orig.head(5)

Unnamed: 0                        id             connectionTime  \
0           0  5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1           1  5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2           2  5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   
3           3  5e23b149f9af8b5fe4b973d2  2020-01-02 13:59:58+00:00   
4           4  5e23b149f9af8b5fe4b973d3  2020-01-02 14:00:01+00:00   

              disconnectTime           doneChargingTime  kWhDelivered  \
0  2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1  2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2  2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   
3  2020-01-02 16:38:39+00:00  2020-01-02 15:18:45+00:00         2.355   
4  2020-01-02 22:08:40+00:00  2020-01-02 18:17:30+00:00        13.375   

                                sessionID  siteID  spaceID    stationID  \
0  1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1  1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2  1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3  1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4  1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   

              timezone  userID  \
0  America/Los_Angeles   194.0   
1  America/Los_Angeles  4275.0   
2  America/Los_Angeles   344.0   
3  America/Los_Angeles  1117.0   
4  America/Los_Angeles   334.0   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                     userInputs  
0                                                                                                                                                                                                                                      [{'WhPerMile': 250, 'kWhRequested': 25.0, 'milesRequested': 100, 'minutesAvailable': 463, 'modifiedAt': 'Thu, 02 Jan 2020 13:09:39 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 20:51:54 GMT', 'userID': 194}]  
1                                                                                                                                                                                                                                     [{'WhPerMile': 280, 'kWhRequested': 70.0, 'milesRequested': 250, 'minutesAvailable': 595, 'modifiedAt': 'Thu, 02 Jan 2020 13:37:11 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 23:31:50 GMT', 'userID': 4275}]  
2                                                                                                                                                                                                                                         [{'WhPerMile': 400, 'kWhRequested': 8.0, 'milesRequested': 20, 'minutesAvailable': 60, 'modifiedAt': 'Thu, 02 Jan 2020 13:57:17 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 14:56:35 GMT', 'userID': 344}]  
3  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'milesRequested': 20, 'minutesAvailable': 65, 'modifiedAt': 'Thu, 02 Jan 2020 14:00:03 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 15:04:58 GMT', 'userID': 1117}, {'WhPerMile': 400, 'kWhRequested': 8.0, 'milesRequested': 20, 'minutesAvailable': 65, 'modifiedAt': 'Thu, 02 Jan 2020 14:00:19 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 15:04:58 GMT', 'userID': 1117}]  
4                                                                                                                                                            

In [5]:
charging_sessions_orig.nunique()

Unnamed: 0          15292
id                  65037
connectionTime      64839
disconnectTime      64906
doneChargingTime    60637
kWhDelivered        25629
sessionID           65037
siteID                  2
spaceID               107
stationID             107
timezone                1
userID               1006
userInputs          47822
dtype: int64

Judging by the number of unique id's in comparison to the whole dataset, we have duplicates. Let's drop them

In [6]:
charging_sessions_modified = charging_sessions_orig.drop_duplicates()

In [7]:
charging_sessions_modified.info()
charging_sessions_modified.nunique()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 66450 entries, 0 to 66449
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        66450 non-null  int64  
 1   id                66450 non-null  object 
 2   connectionTime    66450 non-null  object 
 3   disconnectTime    66450 non-null  object 
 4   doneChargingTime  62362 non-null  object 
 5   kWhDelivered      66450 non-null  float64
 6   sessionID         66450 non-null  object 
 7   siteID            66450 non-null  int64  
 8   spaceID           66450 non-null  object 
 9   stationID         66450 non-null  object 
 10  timezone          66450 non-null  object 
 11  userID            49187 non-null  float64
 12  userInputs        49187 non-null  object 
dtypes: float64(2), int64(2), object(9)
memory usage: 7.1+ MB


Unnamed: 0          15292
id                  65037
connectionTime      64839
disconnectTime      64906
doneChargingTime    60637
kWhDelivered        25629
sessionID           65037
siteID                  2
spaceID               107
stationID             107
timezone                1
userID               1006
userInputs          47822
dtype: int64

This didnt quite do the job, we still have the same amount of entries in the dataset. We need to find out where they differ. But first, let's drop the timezone, since its the same for all of the data.
We also have no information about the the Unnamed: 0 column. It looks to be some kind of identifier, but doesn't have enough unique values to uniquely identify all data points. Since we have no further information, assume this is a data fragment of whoever created the dataset, that we weren't supposed to get. We will drop it too.  

In [8]:
charging_sessions_modified = charging_sessions_orig.drop(columns = ["Unnamed: 0","timezone"])

In [9]:
charging_sessions_modified.info()
charging_sessions_modified.nunique()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66450 entries, 0 to 66449
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                66450 non-null  object 
 1   connectionTime    66450 non-null  object 
 2   disconnectTime    66450 non-null  object 
 3   doneChargingTime  62362 non-null  object 
 4   kWhDelivered      66450 non-null  float64
 5   sessionID         66450 non-null  object 
 6   siteID            66450 non-null  int64  
 7   spaceID           66450 non-null  object 
 8   stationID         66450 non-null  object 
 9   userID            49187 non-null  float64
 10  userInputs        49187 non-null  object 
dtypes: float64(2), int64(1), object(8)
memory usage: 5.6+ MB


id                  65037
connectionTime      64839
disconnectTime      64906
doneChargingTime    60637
kWhDelivered        25629
sessionID           65037
siteID                  2
spaceID               107
stationID             107
userID               1006
userInputs          47822
dtype: int64

In [10]:
# Filter out all rows where id is a duplicate
duplicates = charging_sessions_modified[charging_sessions_modified.duplicated("id", keep = False)]

In [11]:
duplicates.sort_values(by = "sessionID")

id             connectionTime  \
65046  5d310f54f9af8b52adda3e28  2019-07-02 13:25:01+00:00   
5582   5d310f54f9af8b52adda3e28  2019-07-02 13:25:01+00:00   
65124  5d3260aff9af8b56e4697230  2019-07-03 16:20:04+00:00   
5660   5d3260aff9af8b56e4697230  2019-07-03 16:20:04+00:00   
5694   5d38f852f9af8b69fe3549fd  2019-07-08 13:09:26+00:00   
...                         ...                        ...   
66320  5d55f950f9af8b45dfb3d875  2019-07-30 11:31:17+00:00   
66382  5d55f950f9af8b45dfb3d8b3  2019-07-30 22:30:46+00:00   
6918   5d55f950f9af8b45dfb3d8b3  2019-07-30 22:30:46+00:00   
66393  5d574ad2f9af8b4c10c0361e  2019-07-31 11:57:10+00:00   
6929   5d574ad2f9af8b4c10c0361e  2019-07-31 11:57:10+00:00   

                  disconnectTime           doneChargingTime  kWhDelivered  \
65046  2019-07-02 22:39:18+00:00  2019-07-02 16:29:23+00:00        13.924   
5582   2019-07-02 22:39:18+00:00  2019-07-02 16:29:23+00:00        13.924   
65124  2019-07-03 22:41:16+00:00  2019-07-03 20:37:22+00:00        11.184   
5660   2019-07-03 22:41:16+00:00  2019-07-03 20:37:22+00:00        11.184   
5694   2019-07-08 19:03:10+00:00  2019-07-08 17:49:27+00:00        11.093   
...                          ...                        ...           ...   
66320  2019-07-30 22:07:15+00:00  2019-07-30 20:57:54+00:00        46.863   
66382  2019-07-31 00:44:02+00:00  2019-07-31 00:45:01+00:00         4.166   
6918   2019-07-31 00:44:02+00:00  2019-07-31 00:45:01+00:00         4.166   
66393  2019-07-31 21:20:17+00:00  2019-07-31 19:08:13+00:00        25.971   
6929   2019-07-31 21:20:17+00:00  2019-07-31 19:08:13+00:00        25.971   

                                    sessionID  siteID  spaceID    stationID  \
65046  1_1_178_817_2019-07-02 13:25:01.464654       1  AG-1F09  1-1-178-817   
5582   1_1_178_817_2019-07-02 13:25:01.464654       1  AG-1F09  1-1-178-817   
65124  1_1_178_817_2019-07-03 16:20:03.809665       1  AG-1F09  1-1-178-817   
5660   1_1_178_817_2019-07-03 16:20:03.809665       1  AG-1F09  1-1-178-817   
5694   1_1_178_817_2019-07-08 13:08:13.199752       1  AG-1F09  1-1-178-817   
...                                       ...     ...      ...          ...   
66320  1_1_194_826_2019-07-30 11:31:16.894598       1  AG-1F11  1-1-194-826   
66382  1_1_194_826_2019-07-30 22:30:45.743815       1  AG-1F11  1-1-194-826   
6918   1_1_194_826_2019-07-30 22:30:45.743815       1  AG-1F11  1-1-194-826   
66393  1_1_194_826_2019-07-31 11:57:09.786222       1  AG-1F11  1-1-194-826   
6929   1_1_194_826_2019-07-31 11:57:09.786222       1  AG-1F11  1-1-194-826   

       userID  \
65046   322.0   
5582    322.0   
65124   357.0   
5660    357.0   
5694   2500.0   
...       ...   
66320   933.0   
66382    76.0   
6918     76.0   
66393   933.0   
6929    933.0   

                                                                                                                                                                                                                                     userInputs  
65046  [{'WhPerMile': 313, 'kWhRequested': 18.78, 'milesRequested': 60, 'minutesAvailable': 418, 'modifiedAt': 'Tue, 02 Jul 2019 13:25:07 GMT', 'paymentRequired': True, 'requestedDeparture': 'Tue, 02 Jul 2019 20:23:01 GMT', 'userID': 322}]  
5582   [{'WhPerMile': 313, 'kWhRequested': 18.78, 'milesRequested': 60, 'minutesAvailable': 418, 'modifiedAt': 'Tue, 02 Jul 2019 13:25:07 GMT', 'paymentRequired': True, 'requestedDeparture': 'Tue, 02 Jul 2019 20:23:01 GMT', 'userID': 322}]  
65124  [{'WhPerMile': 354, 'kWhRequested': 14.16, 'milesRequested': 40, 'minutesAvailable': 238, 'modifiedAt': 'Wed, 03 Jul 2019 16:21:04 GMT', 'paymentRequired': True, 'requestedDeparture': 'Wed, 03 Jul 2019 20:18:04 GMT', 'userID': 357}]  
5660   [{'WhPerMile': 354, 'kWhRequested': 14.16, 'milesRequested': 40, 'minutesAvailable': 238, 'modifiedAt': 'Wed, 03 Jul 2019 16:21:04 GMT', 'paymentRequired': True, 'requestedDeparture': 'Wed, 03 Jul 2019 20:18:04 GMT', 'userID': 357

The duplicate rows look exactly the same, let's try dropping duplicates again.

In [12]:
charging_sessions_modified.drop_duplicates(inplace = True)

In [13]:
charging_sessions_modified.info()
charging_sessions_modified.nunique()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 65037 entries, 0 to 65036
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                65037 non-null  object 
 1   connectionTime    65037 non-null  object 
 2   disconnectTime    65037 non-null  object 
 3   doneChargingTime  60950 non-null  object 
 4   kWhDelivered      65037 non-null  float64
 5   sessionID         65037 non-null  object 
 6   siteID            65037 non-null  int64  
 7   spaceID           65037 non-null  object 
 8   stationID         65037 non-null  object 
 9   userID            47822 non-null  float64
 10  userInputs        47822 non-null  object 
dtypes: float64(2), int64(1), object(8)
memory usage: 6.0+ MB


id                  65037
connectionTime      64839
disconnectTime      64906
doneChargingTime    60637
kWhDelivered        25629
sessionID           65037
siteID                  2
spaceID               107
stationID             107
userID               1006
userInputs          47822
dtype: int64

That did the job, we now have the same amount of unique id's as we have total rows.
We can now safely drop the id column. Since the sessionID is an aggregate of the connection time and station id, we will drop it too. 

In [14]:
charging_sessions_modified.drop(columns = ["id", "sessionID"], inplace = True)

In [15]:
charging_sessions_modified.head()

connectionTime             disconnectTime  \
0  2020-01-02 13:08:54+00:00  2020-01-02 19:11:15+00:00   
1  2020-01-02 13:36:50+00:00  2020-01-02 22:38:21+00:00   
2  2020-01-02 13:56:35+00:00  2020-01-03 00:39:22+00:00   
3  2020-01-02 13:59:58+00:00  2020-01-02 16:38:39+00:00   
4  2020-01-02 14:00:01+00:00  2020-01-02 22:08:40+00:00   

            doneChargingTime  kWhDelivered  siteID  spaceID    stationID  \
0  2020-01-02 17:31:35+00:00        25.016       1  AG-3F30  1-1-179-810   
1  2020-01-02 20:18:05+00:00        33.097       1  AG-1F01  1-1-193-825   
2  2020-01-02 16:35:06+00:00         6.521       1  AG-1F03  1-1-193-829   
3  2020-01-02 15:18:45+00:00         2.355       1  AG-1F04  1-1-193-820   
4  2020-01-02 18:17:30+00:00        13.375       1  AG-1F06  1-1-193-819   

   userID  \
0   194.0   
1  4275.0   
2   344.0   
3  1117.0   
4   334.0   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                     userInputs  
0                                                                                                                                                                                                                                      [{'WhPerMile': 250, 'kWhRequested': 25.0, 'milesRequested': 100, 'minutesAvailable': 463, 'modifiedAt': 'Thu, 02 Jan 2020 13:09:39 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 20:51:54 GMT', 'userID': 194}]  
1                                                                                                                                                                                                                                     [{'WhPerMile': 280, 'kWhRequested': 70.0, 'milesRequested': 250, 'minutesAvailable': 595, 'modifiedAt': 'Thu, 02 Jan 2020 13:37:11 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 23:31:50 GMT', 'userID': 4275}]  
2                                                                                                                                                                                                                                         [{'WhPerMile': 400, 'kWhRequested': 8.0, 'milesRequested': 20, 'minutesAvailable': 60, 'modifiedAt': 'Thu, 02 Jan 2020 13:57:17 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 14:56:35 GMT', 'userID': 344}]  
3  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'milesRequested': 20, 'minutesAvailable': 65, 'modifiedAt': 'Thu, 02 Jan 2020 14:00:03 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 15:04:58 GMT', 'userID': 1117}, {'WhPerMile': 400, 'kWhRequested': 8.0, 'milesRequested': 20, 'minutesAvailable': 65, 'modifiedAt': 'Thu, 02 Jan 2020 14:00:19 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 15:04:58 GMT', 'userID': 1117}]  
4                                                                                                                                                                                                                                       [{'WhPerMile': 400, 'kWhRequested': 16.0, 'milesRequested': 40, 'minutesAvailable': 504, 'modifiedAt': 'Thu, 02 Jan 2020 14:00:13 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 22:24:01 GMT', 'userID': 334}]

In the last nunique() call, both spaceID and stationID have the same amount of unique data points. Since one identifies the parking space and the other the charging station, it is possible that every parking space has their own charging station, as opposed to one station servicing multiple spaces. If we can prove this, we could drop one of the two columns, since it is redundant.

In [16]:
#Create an aggregate column that contains the concatenated value of spaceID and stationID
charging_sessions_modified["combinedSpaceStationID"] = charging_sessions_modified["spaceID"].astype(str) \
                                                       + "/" + charging_sessions_modified["stationID"].astype(str)
charging_sessions_modified["combinedSpaceStationID"].nunique()

107

The combined column has the same amount of unique values, as the spaceID and stationID. Our assumption holds and we will thus drop the two columns and keep the aggregate. 

In [17]:
charging_sessions_modified.drop(columns = ["spaceID", "stationID"], inplace = True)

Convert the timestamp strings to datetime for easier conversion later on.

In [18]:
charging_sessions_modified["connectionTime"] = pd.to_datetime(charging_sessions_modified["connectionTime"])
charging_sessions_modified["disconnectTime"] = pd.to_datetime(charging_sessions_modified["disconnectTime"])
charging_sessions_modified["doneChargingTime"] = pd.to_datetime(charging_sessions_modified["doneChargingTime"],errors = "coerce")

In [19]:
charging_sessions_modified.head()

connectionTime            disconnectTime  \
0 2020-01-02 13:08:54+00:00 2020-01-02 19:11:15+00:00   
1 2020-01-02 13:36:50+00:00 2020-01-02 22:38:21+00:00   
2 2020-01-02 13:56:35+00:00 2020-01-03 00:39:22+00:00   
3 2020-01-02 13:59:58+00:00 2020-01-02 16:38:39+00:00   
4 2020-01-02 14:00:01+00:00 2020-01-02 22:08:40+00:00   

           doneChargingTime  kWhDelivered  siteID  userID  \
0 2020-01-02 17:31:35+00:00        25.016       1   194.0   
1 2020-01-02 20:18:05+00:00        33.097       1  4275.0   
2 2020-01-02 16:35:06+00:00         6.521       1   344.0   
3 2020-01-02 15:18:45+00:00         2.355       1  1117.0   
4 2020-01-02 18:17:30+00:00        13.375       1   334.0   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                     userInputs  \
0                                                                                                                                                                                                                                      [{'WhPerMile': 250, 'kWhRequested': 25.0, 'milesRequested': 100, 'minutesAvailable': 463, 'modifiedAt': 'Thu, 02 Jan 2020 13:09:39 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 20:51:54 GMT', 'userID': 194}]   
1                                                                                                                                                                                                                                     [{'WhPerMile': 280, 'kWhRequested': 70.0, 'milesRequested': 250, 'minutesAvailable': 595, 'modifiedAt': 'Thu, 02 Jan 2020 13:37:11 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 23:31:50 GMT', 'userID': 4275}]   
2                                                                                                                                                                                                                                         [{'WhPerMile': 400, 'kWhRequested': 8.0, 'milesRequested': 20, 'minutesAvailable': 60, 'modifiedAt': 'Thu, 02 Jan 2020 13:57:17 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 14:56:35 GMT', 'userID': 344}]   
3  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'milesRequested': 20, 'minutesAvailable': 65, 'modifiedAt': 'Thu, 02 Jan 2020 14:00:03 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 15:04:58 GMT', 'userID': 1117}, {'WhPerMile': 400, 'kWhRequested': 8.0, 'milesRequested': 20, 'minutesAvailable': 65, 'modifiedAt': 'Thu, 02 Jan 2020 14:00:19 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 15:04:58 GMT', 'userID': 1117}]   
4                                                                                                                                                                                                                                       [{'WhPerMile': 400, 'kWhRequested': 16.0, 'milesRequested': 40, 'minutesAvailable': 504, 'modifiedAt': 'Thu, 02 Jan 2020 14:00:13 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 22:24:01 GMT', 'userID': 334}]   

  combinedSpaceStationID  
0    AG-3F30/1-1-179-810  
1    AG-1F01/1-1-193-825  
2    AG-1F03/1-1-193-829  
3    AG-1F04/1-1-193-820  
4    AG-1F06/1-1-193-819

Let's do a check, whether we have any data where the disconnectTime occurs before the connectionTime. Since this is logically impossible, we would drop this data.

In [20]:
charging_sessions_modified[charging_sessions_modified["disconnectTime"]<charging_sessions_modified["connectionTime"]]

Empty DataFrame
Columns: [connectionTime, disconnectTime, doneChargingTime, kWhDelivered, siteID, userID, userInputs, combinedSpaceStationID]
Index: []

Doesn't occur anywhere, we're good on this front.  
At this point we are gonna make an assumption on the dataset. We still have null values in the doneChargingTime column. Our assumption is, these values are null if disconnect and doneChargingTime would be the same, i.e. if the the vehicle gets disconnected before it finished charging. We try to generate some supporting evidence, by checking if the disconnect and doneCharging are the same anywhere in the dataset.

In [21]:
charging_sessions_modified[charging_sessions_modified["disconnectTime"] == charging_sessions_modified["doneChargingTime"]]

connectionTime            disconnectTime  \
25051 2019-05-31 22:26:39+00:00 2019-06-01 00:24:47+00:00   
35096 2018-07-09 00:21:39+00:00 2018-07-09 06:16:09+00:00   
40658 2018-09-20 18:21:06+00:00 2018-09-20 20:20:04+00:00   
41408 2018-09-29 22:03:35+00:00 2018-09-29 23:00:39+00:00   
42038 2018-10-08 00:36:11+00:00 2018-10-08 01:42:18+00:00   

               doneChargingTime  kWhDelivered  siteID  userID  \
25051 2019-06-01 00:24:47+00:00      5.933000       2   431.0   
35096 2018-07-09 06:16:09+00:00     27.154000       2     NaN   
40658 2018-09-20 20:20:04+00:00      3.584000       2     NaN   
41408 2018-09-29 23:00:39+00:00      6.815959       2     NaN   
42038 2018-10-08 01:42:18+00:00      7.191020       2     NaN   

                                                                                                                                                                                                                                    userInputs  \
25051  [{'WhPerMile': 300, 'kWhRequested': 18.0, 'milesRequested': 60, 'minutesAvailable': 149, 'modifiedAt': 'Fri, 31 May 2019 22:27:14 GMT', 'paymentRequired': True, 'requestedDeparture': 'Sat, 01 Jun 2019 00:55:39 GMT', 'userID': 431}]   
35096                                                                                                                                                                                                                                      NaN   
40658                                                                                                                                                                                                                                      NaN   
41408                                                                                                                                                                                                                                      NaN   
42038                                                                                                                                                                                                                                      NaN   

      combinedSpaceStationID  
25051     CA-322/2-39-78-360  
35096     CA-320/2-39-78-363  
40658     CA-323/2-39-78-366  
41408     CA-323/2-39-78-366  
42038     CA-323/2-39-78-366

Because we have a few data points, where the disconnect and doneCharging time are identical, we don't have any evidence to support our earlier assumption. It might still be possible that it holds true, but we will play it safe and drop all the rows where the doneChargingTime is null.

In [22]:
charging_sessions_modified.dropna(axis = 0, subset = "doneChargingTime", inplace=True)

In [23]:
charging_sessions_modified.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 60950 entries, 0 to 65036
Data columns (total 8 columns):
 #   Column                  Non-Null Count  Dtype              
---  ------                  --------------  -----              
 0   connectionTime          60950 non-null  datetime64[ns, UTC]
 1   disconnectTime          60950 non-null  datetime64[ns, UTC]
 2   doneChargingTime        60950 non-null  datetime64[ns, UTC]
 3   kWhDelivered            60950 non-null  float64            
 4   siteID                  60950 non-null  int64              
 5   userID                  44644 non-null  float64            
 6   userInputs              44644 non-null  object             
 7   combinedSpaceStationID  60950 non-null  object             
dtypes: datetime64[ns, UTC](3), float64(2), int64(1), object(2)
memory usage: 4.2+ MB


This leaves us with null-values only in the userID and inputs column. We will deal with these later on.
Before that, we will add a few columns that contain relevant metrics. Starting with the total time the vehicle was connected to the charging station, the time it was actually charging and the idle time (each in seconds).

In [24]:
charging_sessions_modified = charging_sessions_modified.assign(timeConnected = lambda x : ((x["disconnectTime"]-x["connectionTime"])))
charging_sessions_modified["timeConnected"] = charging_sessions_modified["timeConnected"].dt.total_seconds()

charging_sessions_modified = charging_sessions_modified.assign(idleTime = lambda x : ((x["disconnectTime"]-x["doneChargingTime"])))
charging_sessions_modified["idleTime"] = charging_sessions_modified["idleTime"].dt.total_seconds()

charging_sessions_modified = charging_sessions_modified.assign(chargingTime = lambda x : ((x["doneChargingTime"]-x["connectionTime"])))
charging_sessions_modified["chargingTime"] = charging_sessions_modified["chargingTime"].dt.total_seconds()

In [25]:
charging_sessions_modified.head()

connectionTime            disconnectTime  \
0 2020-01-02 13:08:54+00:00 2020-01-02 19:11:15+00:00   
1 2020-01-02 13:36:50+00:00 2020-01-02 22:38:21+00:00   
2 2020-01-02 13:56:35+00:00 2020-01-03 00:39:22+00:00   
3 2020-01-02 13:59:58+00:00 2020-01-02 16:38:39+00:00   
4 2020-01-02 14:00:01+00:00 2020-01-02 22:08:40+00:00   

           doneChargingTime  kWhDelivered  siteID  userID  \
0 2020-01-02 17:31:35+00:00        25.016       1   194.0   
1 2020-01-02 20:18:05+00:00        33.097       1  4275.0   
2 2020-01-02 16:35:06+00:00         6.521       1   344.0   
3 2020-01-02 15:18:45+00:00         2.355       1  1117.0   
4 2020-01-02 18:17:30+00:00        13.375       1   334.0   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                     userInputs  \
0                                                                                                                                                                                                                                      [{'WhPerMile': 250, 'kWhRequested': 25.0, 'milesRequested': 100, 'minutesAvailable': 463, 'modifiedAt': 'Thu, 02 Jan 2020 13:09:39 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 20:51:54 GMT', 'userID': 194}]   
1                                                                                                                                                                                                                                     [{'WhPerMile': 280, 'kWhRequested': 70.0, 'milesRequested': 250, 'minutesAvailable': 595, 'modifiedAt': 'Thu, 02 Jan 2020 13:37:11 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 23:31:50 GMT', 'userID': 4275}]   
2                                                                                                                                                                                                                                         [{'WhPerMile': 400, 'kWhRequested': 8.0, 'milesRequested': 20, 'minutesAvailable': 60, 'modifiedAt': 'Thu, 02 Jan 2020 13:57:17 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 14:56:35 GMT', 'userID': 344}]   
3  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'milesRequested': 20, 'minutesAvailable': 65, 'modifiedAt': 'Thu, 02 Jan 2020 14:00:03 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 15:04:58 GMT', 'userID': 1117}, {'WhPerMile': 400, 'kWhRequested': 8.0, 'milesRequested': 20, 'minutesAvailable': 65, 'modifiedAt': 'Thu, 02 Jan 2020 14:00:19 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 15:04:58 GMT', 'userID': 1117}]   
4                                                                                                                                                                                                                                       [{'WhPerMile': 400, 'kWhRequested': 16.0, 'milesRequested': 40, 'minutesAvailable': 504, 'modifiedAt': 'Thu, 02 Jan 2020 14:00:13 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 22:24:01 GMT', 'userID': 334}]   

  combinedSpaceStationID  timeConnected  idleTime  chargingTime  
0    AG-3F30/1-1-179-810        21741.0    5980.0       15761.0  
1    AG-1F01/1-1-193-825        32491.0    8416.0       24075.0  
2    AG-1F03/1-1-193-829        38567.0   29056.0        9511.0  
3    AG-1F04/1-1-193-820         9521.0    4794.0        4727.0  
4    AG-1F06/1-1-193-819        29319.0   13870.0       15449.0

Other useful columns that we will now add for seasonality and temporal pattern analysis are the day of the week, the time and the month

In [26]:
charging_sessions_modified["weekday"] = charging_sessions_modified["connectionTime"].dt.day_of_week
charging_sessions_modified["month"] = charging_sessions_modified["connectionTime"].dt.month
charging_sessions_modified["time"] = charging_sessions_modified["connectionTime"].dt.time
charging_sessions_modified.head()

connectionTime            disconnectTime  \
0 2020-01-02 13:08:54+00:00 2020-01-02 19:11:15+00:00   
1 2020-01-02 13:36:50+00:00 2020-01-02 22:38:21+00:00   
2 2020-01-02 13:56:35+00:00 2020-01-03 00:39:22+00:00   
3 2020-01-02 13:59:58+00:00 2020-01-02 16:38:39+00:00   
4 2020-01-02 14:00:01+00:00 2020-01-02 22:08:40+00:00   

           doneChargingTime  kWhDelivered  siteID  userID  \
0 2020-01-02 17:31:35+00:00        25.016       1   194.0   
1 2020-01-02 20:18:05+00:00        33.097       1  4275.0   
2 2020-01-02 16:35:06+00:00         6.521       1   344.0   
3 2020-01-02 15:18:45+00:00         2.355       1  1117.0   
4 2020-01-02 18:17:30+00:00        13.375       1   334.0   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                     userInputs  \
0                                                                                                                                                                                                                                      [{'WhPerMile': 250, 'kWhRequested': 25.0, 'milesRequested': 100, 'minutesAvailable': 463, 'modifiedAt': 'Thu, 02 Jan 2020 13:09:39 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 20:51:54 GMT', 'userID': 194}]   
1                                                                                                                                                                                                                                     [{'WhPerMile': 280, 'kWhRequested': 70.0, 'milesRequested': 250, 'minutesAvailable': 595, 'modifiedAt': 'Thu, 02 Jan 2020 13:37:11 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 23:31:50 GMT', 'userID': 4275}]   
2                                                                                                                                                                                                                                         [{'WhPerMile': 400, 'kWhRequested': 8.0, 'milesRequested': 20, 'minutesAvailable': 60, 'modifiedAt': 'Thu, 02 Jan 2020 13:57:17 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 14:56:35 GMT', 'userID': 344}]   
3  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'milesRequested': 20, 'minutesAvailable': 65, 'modifiedAt': 'Thu, 02 Jan 2020 14:00:03 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 15:04:58 GMT', 'userID': 1117}, {'WhPerMile': 400, 'kWhRequested': 8.0, 'milesRequested': 20, 'minutesAvailable': 65, 'modifiedAt': 'Thu, 02 Jan 2020 14:00:19 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 15:04:58 GMT', 'userID': 1117}]   
4                                                                                                                                                                                                                                       [{'WhPerMile': 400, 'kWhRequested': 16.0, 'milesRequested': 40, 'minutesAvailable': 504, 'modifiedAt': 'Thu, 02 Jan 2020 14:00:13 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 22:24:01 GMT', 'userID': 334}]   

  combinedSpaceStationID  timeConnected  idleTime  chargingTime  weekday  \
0    AG-3F30/1-1-179-810        21741.0    5980.0       15761.0        3   
1    AG-1F01/1-1-193-825        32491.0    8416.0       24075.0        3   
2    AG-1F03/1-1-193-829        38567.0   29056.0        9511.0        3   
3    AG-1F04/1-1-193-820         9521.0    4794.0        4727.0        3   
4    AG-1F06/1-1-193-819        29319.0   13870.0       15449.0        3   

   month      time  
0      1  13:08:54  
1

Now we will check for erroneous data points.

In [27]:
charging_sessions_modified.describe().transpose()

count          mean           std       min        25%  \
kWhDelivered   60950.0     11.583268     10.153601     0.501      5.013   
siteID         60950.0      1.481674      0.499668     1.000      1.000   
userID         44644.0   1709.660514   2464.480169     1.000    431.000   
timeConnected  60950.0  22960.401198  17717.646112   183.000  10708.500   
idleTime       60950.0  10271.909319  14447.964854 -3596.000     13.000   
chargingTime   60950.0  12688.491879  11032.329226 -2482.000   5823.250   
weekday        60950.0      2.281920      1.650711     0.000      1.000   
month          60950.0      6.596702      3.370832     1.000      4.000   

                     50%        75%            max  
kWhDelivered       8.991     13.984     108.797242  
siteID             1.000      2.000       2.000000  
userID           774.000   1754.000   19923.000000  
timeConnected  22705.000  33267.000  882969.000000  
idleTime        5774.000  17712.750  562037.000000  
chargingTime    9933.000  17011.000  720057.000000  
weekday            2.000      4.000       6.000000  
month              7.000      9.000      12.000000

There are definitely a few outliers we need to take care off. Negative values in idle and charging time make no sense logically, we will evaluate these to see if its erroneous data or another error we might have introduced.

In [28]:
charging_sessions_modified[charging_sessions_modified["idleTime"]<0]

connectionTime            disconnectTime  \
12    2020-01-02 15:04:38+00:00 2020-01-02 22:08:39+00:00   
20    2020-01-02 15:28:47+00:00 2020-01-02 19:01:54+00:00   
25    2020-01-02 15:42:05+00:00 2020-01-02 21:58:45+00:00   
26    2020-01-02 15:57:24+00:00 2020-01-02 16:35:37+00:00   
33    2020-01-02 16:34:35+00:00 2020-01-02 18:49:41+00:00   
...                         ...                       ...   
65028 2019-07-01 19:20:31+00:00 2019-07-02 00:16:32+00:00   
65030 2019-07-01 20:01:07+00:00 2019-07-02 00:32:26+00:00   
65033 2019-07-01 21:58:45+00:00 2019-07-02 00:39:48+00:00   
65034 2019-07-01 22:02:21+00:00 2019-07-02 00:58:50+00:00   
65035 2019-07-01 22:23:44+00:00 2019-07-02 02:03:36+00:00   

               doneChargingTime  kWhDelivered  siteID  userID  \
12    2020-01-02 22:09:36+00:00        25.567       1   528.0   
20    2020-01-02 19:02:51+00:00         7.417       1  1283.0   
25    2020-01-02 21:59:42+00:00        36.701       1   474.0   
26    2020-01-02 16:36:34+00:00         3.689       1   724.0   
33    2020-01-02 18:50:38+00:00         7.120       1  2276.0   
...                         ...           ...     ...     ...   
65028 2019-07-02 00:16:42+00:00        26.324       1   458.0   
65030 2019-07-02 00:32:59+00:00        21.588       1  1479.0   
65033 2019-07-02 00:40:21+00:00        16.864       1   364.0   
65034 2019-07-02 00:59:23+00:00        18.335       1  2050.0   
65035 2019-07-02 02:04:06+00:00        22.815       1  1626.0   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        userInputs  \
12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        [{'WhPerMile': 250, 'kWhRequested': 50.0, 'milesRequested': 200, 'minutesAvailable': 555, 'modifiedAt': 'Thu, 02 Jan 2020 15:04:49 GMT', 'paymentRequired': True, 'requestedDeparture': 'Fri, 03 Jan 2020 00:19:38 GMT', 'userID': 528}]   
20                                                                                                                                                                                                                                              [{'WhPerMile': 350, 'kWhRequested': 42.0, 'milesRequested': 120, 'minutesAvailable': 238, 'modifiedAt': 'Thu, 02 Jan 2020 15:29:16 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 19:26:47 GMT', 'userID': 1283}, {'WhPerMile': 350, 'kWhRequested': 56.0, 'milesRequested': 160, 'minutesAvailable': 282, 'modifiedAt': 'Thu, 02 Jan 2020 16:29:35 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 20:10:47 GMT', 'userID': 1283}]   
25                                                                                                                                                                                                                                                                                                                                                                                   

In [29]:
charging_sessions_modified[charging_sessions_modified["idleTime"]<0].describe().transpose()

count          mean           std       min       25%  \
kWhDelivered   4387.0     14.239556     12.575748     0.508     5.379   
siteID         4387.0      1.268977      0.443478     1.000     1.000   
userID         3329.0   1458.393812   1532.613614    22.000   410.000   
timeConnected  4387.0  15629.237520  14617.364141   242.000  6534.500   
idleTime       4387.0    -56.017096     76.290027 -3596.000   -59.000   
chargingTime   4387.0  15685.254616  14615.972474   253.000  6592.500   
weekday        4387.0      2.247322      1.660041     0.000     1.000   
month          4387.0      7.178938      3.401890     1.000     5.000   

                     50%         75%            max  
kWhDelivered      11.161     19.2395     108.797242  
siteID             1.000      2.0000       2.000000  
userID           832.000   2052.0000   17097.000000  
timeConnected  12291.000  21362.5000  540779.000000  
idleTime         -58.000    -57.0000      -1.000000  
chargingTime   12351.000  21417.0000  540838.000000  
weekday            2.000      4.0000       6.000000  
month              8.000     10.0000      12.000000

# TODO: Deal with the negative idle time values by ...
Mean is around a minute, keep these, throw out everything else?

Another potential problem are the max values for timeConnected, idleTime and chargingTime. While it is possible to be connected and charging for the up to 10 days the max values represent, it is not that realistic. Let's have a closer look at all data points, where the connection time is above 24 hours (86400 seconds)

In [30]:
charging_sessions_modified[charging_sessions_modified["timeConnected"]>=86400]

connectionTime            disconnectTime  \
7627  2019-08-14 13:00:40+00:00 2019-08-16 02:24:39+00:00   
7911  2019-08-20 14:20:30+00:00 2019-08-22 00:32:21+00:00   
18601 2021-07-16 14:01:18+00:00 2021-07-19 23:05:55+00:00   
18603 2021-07-16 15:39:02+00:00 2021-07-19 23:05:38+00:00   
18604 2021-07-16 15:45:30+00:00 2021-07-19 23:06:05+00:00   
...                         ...                       ...   
54521 2018-11-12 19:35:55+00:00 2018-11-15 01:44:47+00:00   
61169 2019-04-16 00:01:28+00:00 2019-04-18 00:20:42+00:00   
62402 2019-05-09 00:09:47+00:00 2019-05-10 01:23:16+00:00   
63029 2019-05-21 00:10:00+00:00 2019-05-22 00:49:28+00:00   
64086 2019-06-12 13:38:35+00:00 2019-06-14 13:28:08+00:00   

               doneChargingTime  kWhDelivered  siteID   userID  \
7627  2019-08-14 14:48:08+00:00         4.557       1    521.0   
7911  2019-08-20 17:55:00+00:00        14.022       1   2187.0   
18601 2021-07-16 18:16:51+00:00        11.286       1    382.0   
18603 2021-07-16 18:16:54+00:00        13.377       1  10747.0   
18604 2021-07-19 20:36:25+00:00         5.793       1   3161.0   
...                         ...           ...     ...      ...   
54521 2018-11-12 23:41:28+00:00        10.646       1    425.0   
61169 2019-04-17 01:18:18+00:00        39.970       1   2050.0   
62402 2019-05-09 06:04:05+00:00        36.682       1   2050.0   
63029 2019-05-21 23:22:59+00:00        39.989       1   2050.0   
64086 2019-06-12 17:04:30+00:00         6.449       1   2447.0   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  userInputs  \
7627                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  [{'WhPerMile': 400, 'kWhRequested': 16.0, 'milesRequested': 40, 'minutesAvailable': 73, 'modifiedAt': 'Wed, 14 Aug 2019 13:00:41 GMT', 'paymentRequired': True, 'requestedDeparture': 'Wed, 14 Aug 2019 14:13:40 GMT', 'userID': 521}]   
7911                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       

In [31]:
charging_sessions_modified[charging_sessions_modified["timeConnected"]>=86400].describe().transpose()

count           mean            std        min          25%  \
kWhDelivered   271.0      15.776274      13.258068      0.855       6.5795   
siteID         271.0       1.933579       0.249477      1.000       2.0000   
userID         126.0    1900.095238    2378.863569     67.000     565.0000   
timeConnected  271.0  170722.863469  105132.493138  86495.000  101319.0000   
idleTime       271.0  118105.793358   94392.426037    -59.000   67702.0000   
chargingTime   271.0   52617.070111   84529.119803    -59.000    8030.5000   
weekday        271.0       3.173432       2.042918      0.000       1.0000   
month          271.0       6.749077       2.936953      1.000       5.0000   

                      50%         75%         max  
kWhDelivered       13.198      15.856      61.114  
siteID              2.000       2.000       2.000  
userID           1202.000    1826.000   15803.000  
timeConnected  134603.000  200845.500  882969.000  
idleTime        93320.000  150709.000  562037.000  
chargingTime    17079.000   75423.000  720057.000  
weekday             4.000       5.000       6.000  
month               7.000       9.000      12.000

271 data points where the connection time is more than a day is quite a lot. The mean of these being around 2 days is also higher than we would have expected.
# TODO: We will deal with these by ...

Now let's check the userInputs column. Since we have null values over more than a fourth of the data-set there, simply dropping them might not be the best solution. Let's have a closer look at the data to see if they are of any use to us.

In [32]:
charging_sessions_modified.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 60950 entries, 0 to 65036
Data columns (total 14 columns):
 #   Column                  Non-Null Count  Dtype              
---  ------                  --------------  -----              
 0   connectionTime          60950 non-null  datetime64[ns, UTC]
 1   disconnectTime          60950 non-null  datetime64[ns, UTC]
 2   doneChargingTime        60950 non-null  datetime64[ns, UTC]
 3   kWhDelivered            60950 non-null  float64            
 4   siteID                  60950 non-null  int64              
 5   userID                  44644 non-null  float64            
 6   userInputs              44644 non-null  object             
 7   combinedSpaceStationID  60950 non-null  object             
 8   timeConnected           60950 non-null  float64            
 9   idleTime                60950 non-null  float64            
 10  chargingTime            60950 non-null  float64            
 11  weekday                 60950 non-null  i

In [33]:
userInputs = pd.DataFrame(charging_sessions_modified.copy()["userInputs"])
userInputs

userInputs
0                                                                                                                                                                                                                                                [{'WhPerMile': 250, 'kWhRequested': 25.0, 'milesRequested': 100, 'minutesAvailable': 463, 'modifiedAt': 'Thu, 02 Jan 2020 13:09:39 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 20:51:54 GMT', 'userID': 194}]
1                                                                                                                                                                                                                                               [{'WhPerMile': 280, 'kWhRequested': 70.0, 'milesRequested': 250, 'minutesAvailable': 595, 'modifiedAt': 'Thu, 02 Jan 2020 13:37:11 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 23:31:50 GMT', 'userID': 4275}]
2                                                                                                                                                                                                                                                   [{'WhPerMile': 400, 'kWhRequested': 8.0, 'milesRequested': 20, 'minutesAvailable': 60, 'modifiedAt': 'Thu, 02 Jan 2020 13:57:17 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 14:56:35 GMT', 'userID': 344}]
3            [{'WhPerMile': 400, 'kWhRequested': 8.0, 'milesRequested': 20, 'minutesAvailable': 65, 'modifiedAt': 'Thu, 02 Jan 2020 14:00:03 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 15:04:58 GMT', 'userID': 1117}, {'WhPerMile': 400, 'kWhRequested': 8.0, 'milesRequested': 20, 'minutesAvailable': 65, 'modifiedAt': 'Thu, 02 Jan 2020 14:00:19 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 15:04:58 GMT', 'userID': 1117}]
4                                                                                                                                                                                                                                                 [{'WhPerMile': 400, 'kWhRequested': 16.0, 'milesRequested': 40, 'minutesAvailable': 504, 'modifiedAt': 'Thu, 02 Jan 2020 14:00:13 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 22:24:01 GMT', 'userID': 334}]
...                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   ...
65032      [{'WhPerMile': 400, 'kWhRequested': 32.0, 'milesRequested': 80, 'minutesAvailable': 36, 'modifiedAt': 'Mon, 01 Jul 2019 21:49:37 GMT', 'paymentRequired': True, 'requestedDeparture': 'Mon, 01 Jul 2019 22:25:12 GMT', 'userID': 1346}, {'WhPerMile': 400, 'kWhRequested': 32.0, 'milesRequested': 80, 'minutesAvailable': 36, 'modifiedAt': 'Mon, 01 Jul 2019 21:51:07 GMT', 'paymentRequired': True, 'requestedDeparture': 'Mon, 01 Jul 2019 22:25:12 GMT', 'userID': 1346}]
65033                                                                                                                                                                                                                                            [{'WhPerMile': 400, 'kWhRequested': 40.0, 'milesRequested': 100, 'minutesAvailable': 234, 'modifiedAt': 'Mon, 01 Jul 2019 21:59:09 GMT', 'paymentRequired': True, 'requestedDeparture': 'Tue, 02 Jul 2019 01:52:45 GMT', 'userID': 364}]
65034  [{'WhPerMile': 333, 'kWhRequested': 29.97, 'milesRequested': 90, 'minutesAvailable': 309, 'modifiedAt': 'Mon, 01 Jul 2019 22:02:37 GMT', 'paymentRequired': True, 'requestedDeparture': 'Tue, 

In [34]:
import numpy as np
# Convert string representations of dictionaries to actual dictionaries
userInputs['userInputs'] = userInputs['userInputs'].apply(lambda x: ast.literal_eval(x) if isinstance(x, str) and pd.notnull(x) else x)

# Define the default dictionary for filling None values within each dictionary
default_dict = {'WhPerMile': np.NaN, 'kWhRequested': np.NaN, 'milesRequested': np.NaN,
                'minutesAvailable': np.NaN, 'modifiedAt':  np.NaN, 'paymentRequired': None,
                'requestedDeparture': np.NaN}

# Function to fill None values within a dictionary with the default dictionary
def fill_none_values(entry):
    if isinstance(entry, dict):
        return {key: default_dict[key] if entry.get(key) is None else entry[key] for key in default_dict}
    elif isinstance(entry, list) and entry and isinstance(entry[0], dict):
        return entry[0]
    else:
        return default_dict

# Apply the fill_none_values function to the "userInputs" column
userInputs['userInputs'] = userInputs['userInputs'].apply(fill_none_values)

# Use json_normalize to flatten the nested dictionaries
user_df = json_normalize(userInputs['userInputs'])

In [35]:
user_df

WhPerMile  kWhRequested  milesRequested  minutesAvailable  \
0          250.0         25.00           100.0             463.0   
1          280.0         70.00           250.0             595.0   
2          400.0          8.00            20.0              60.0   
3          400.0          8.00            20.0              65.0   
4          400.0         16.00            40.0             504.0   
...          ...           ...             ...               ...   
60945      400.0         32.00            80.0              36.0   
60946      400.0         40.00           100.0             234.0   
60947      333.0         29.97            90.0             309.0   
60948      200.0         34.00           170.0              99.0   
60949      333.0         16.65            50.0             191.0   

                          modifiedAt paymentRequired  \
0      Thu, 02 Jan 2020 13:09:39 GMT            True   
1      Thu, 02 Jan 2020 13:37:11 GMT            True   
2      Thu, 02 Jan 2020 13:57:17 GMT            True   
3      Thu, 02 Jan 2020 14:00:03 GMT            True   
4      Thu, 02 Jan 2020 14:00:13 GMT            True   
...                              ...             ...   
60945  Mon, 01 Jul 2019 21:49:37 GMT            True   
60946  Mon, 01 Jul 2019 21:59:09 GMT            True   
60947  Mon, 01 Jul 2019 22:02:37 GMT            True   
60948  Mon, 01 Jul 2019 22:23:43 GMT            True   
60949  Mon, 01 Jul 2019 22:36:08 GMT            True   

                  requestedDeparture  userID  
0      Thu, 02 Jan 2020 20:51:54 GMT   194.0  
1      Thu, 02 Jan 2020 23:31:50 GMT  4275.0  
2      Thu, 02 Jan 2020 14:56:35 GMT   344.0  
3      Thu, 02 Jan 2020 15:04:58 GMT  1117.0  
4      Thu, 02 Jan 2020 22:24:01 GMT   334.0  
...                              ...     ...  
60945  Mon, 01 Jul 2019 22:25:12 GMT  1346.0  
60946  Tue, 02 Jul 2019 01:52:45 GMT   364.0  
60947  Tue, 02 Jul 2019 03:11:21 GMT  2050.0  
60948  Tue, 02 Jul 2019 00:02:44 GMT  1626.0  
60949  Tue, 02 Jul 2019 01:47:08 GMT  2299.0  

[60950 rows x 8 columns]

In [36]:
user_df.info()
user_df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60950 entries, 0 to 60949
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   WhPerMile           44644 non-null  float64
 1   kWhRequested        44644 non-null  float64
 2   milesRequested      44644 non-null  float64
 3   minutesAvailable    44644 non-null  float64
 4   modifiedAt          44644 non-null  object 
 5   paymentRequired     44644 non-null  object 
 6   requestedDeparture  44644 non-null  object 
 7   userID              44644 non-null  float64
dtypes: float64(5), object(3)
memory usage: 3.7+ MB


WhPerMile  kWhRequested  milesRequested  minutesAvailable  \
count  44644.000000  44644.000000    44644.000000      44644.000000   
mean     362.668399     24.763407       72.996707        355.158050   
std      102.329159     20.584336       60.547162        194.843595   
min       50.000000      0.000000        0.000000          1.000000   
25%      290.000000     12.000000       30.000000        191.000000   
50%      360.000000     18.540000       50.000000        360.000000   
75%      400.000000     31.300000      100.000000        498.000000   
max     2000.000000    215.320000      480.000000      10062.000000   

             userID  
count  44644.000000  
mean    1709.660514  
std     2464.480169  
min        1.000000  
25%      431.000000  
50%      774.000000  
75%     1754.000000  
max    19923.000000

Combine this with the rest of the data and drop the now redundant userInputs column

In [37]:
user_df.drop(columns=["userID"], inplace=True) #drop this so we dont duplicate it in the concatenated dataframe
charging_sessions_modified = pd.concat([charging_sessions_modified,user_df], axis = 1)
charging_sessions_modified.drop(columns= ["userInputs"], inplace= True )

In [38]:
charging_sessions_modified

connectionTime            disconnectTime  \
0     2020-01-02 13:08:54+00:00 2020-01-02 19:11:15+00:00   
1     2020-01-02 13:36:50+00:00 2020-01-02 22:38:21+00:00   
2     2020-01-02 13:56:35+00:00 2020-01-03 00:39:22+00:00   
3     2020-01-02 13:59:58+00:00 2020-01-02 16:38:39+00:00   
4     2020-01-02 14:00:01+00:00 2020-01-02 22:08:40+00:00   
...                         ...                       ...   
60303                       NaT                       NaT   
60312                       NaT                       NaT   
60326                       NaT                       NaT   
60396                       NaT                       NaT   
60411                       NaT                       NaT   

               doneChargingTime  kWhDelivered  siteID  userID  \
0     2020-01-02 17:31:35+00:00        25.016     1.0   194.0   
1     2020-01-02 20:18:05+00:00        33.097     1.0  4275.0   
2     2020-01-02 16:35:06+00:00         6.521     1.0   344.0   
3     2020-01-02 15:18:45+00:00         2.355     1.0  1117.0   
4     2020-01-02 18:17:30+00:00        13.375     1.0   334.0   
...                         ...           ...     ...     ...   
60303                       NaT           NaN     NaN     NaN   
60312                       NaT           NaN     NaN     NaN   
60326                       NaT           NaN     NaN     NaN   
60396                       NaT           NaN     NaN     NaN   
60411                       NaT           NaN     NaN     NaN   

      combinedSpaceStationID  timeConnected  idleTime  chargingTime  weekday  \
0        AG-3F30/1-1-179-810        21741.0    5980.0       15761.0      3.0   
1        AG-1F01/1-1-193-825        32491.0    8416.0       24075.0      3.0   
2        AG-1F03/1-1-193-829        38567.0   29056.0        9511.0      3.0   
3        AG-1F04/1-1-193-820         9521.0    4794.0        4727.0      3.0   
4        AG-1F06/1-1-193-819        29319.0   13870.0       15449.0      3.0   
...                      ...            ...       ...           ...      ...   
60303                    NaN            NaN       NaN           NaN      NaN   
60312                    NaN            NaN       NaN           NaN      NaN   
60326                    NaN            NaN       NaN           NaN      NaN   
60396                    NaN            NaN       NaN           NaN      NaN   
60411                    NaN            NaN       NaN           NaN      NaN   

       month      time  WhPerMile  kWhRequested  milesRequested  \
0        1.0  13:08:54      250.0         25.00           100.0   
1        1.0  13:36:50      280.0         70.00           250.0   
2        1.0  13:56:35      400.0          8.00            20.0   
3        1.0  13:59:58      400.0          8.00            20.0   
4        1.0  14:00:01      400.0         16.00            40.0   
...      ...       ...        ...           ...             ...   
60303    NaN       NaN      300.0         15.00            50.0   
60312    NaN       NaN      288.0         23.04            80.0   
60326    NaN       NaN      292.0         17.52            60.0   
60396    NaN       NaN      364.0         21.84            60.0   
60411    NaN       NaN      273.0         27.30           100.0   

       minutesAvailable                     modifiedAt paymentRequired  \
0                 463.0  Thu, 02 Jan 2020 13:09:39 GMT            True   
1                 595.0  Thu, 02 Jan 2020 13:37:11 GMT            True   
2                  60.0  Thu, 02 Jan 2020 13:57:17 GMT            True   
3                  65.0  Thu, 02 Jan 2020 14:00:03 GMT            True   
4                 504.0  Thu, 02 Jan 2020 14:00:13 GMT            True   
...                 ...                            ...             ...   
60303             453.0  Tue, 18 Jun 2019 14:05:24 GMT            True   
60312             473.0  Tue, 18 Jun 2019 14:20:49 GMT            True   
60326             169.0  Tue, 18 Jun 2019 22:01:10 GMT            True 

In [39]:
weather_orig.describe().transpose()

count        mean       std    min    25%     50%     75%  \
temperature       29219.0   17.885622  6.454604    2.0   13.0   17.00   22.00   
cloud_cover       29224.0   30.107993  5.400388    4.0   26.0   33.00   34.00   
pressure          29236.0  986.879400  3.607065  971.0  984.5  986.48  989.11   
windspeed         29158.0    8.566568  6.699889    0.0    6.0    7.00   13.00   
precipitation     29244.0    0.059360  0.471682    0.0    0.0    0.00    0.00   
felt_temperature  29218.0   17.731433  6.321100    0.0   13.0   17.00   22.00   

                     max  
temperature        46.00  
cloud_cover        47.00  
pressure          999.65  
windspeed          57.00  
precipitation      18.54  
felt_temperature   42.00